### PR TITLE
C# Fix loading the native library on dotnet cli

### PR DESCRIPTION
**Pull request by jskeet**
Two separate commits:

- Add a property to GrpcEnvironment to allow the user to specify the native library
- Change how the library is loaded and used

### CODE REVIEW

File: src/csharp/Grpc.Core/Internal/UnmanagedLibrary.cs [hunk 1]
```diff
@@ -82,6 +82,32 @@ public UnmanagedLibrary(string[] libraryPathAlternatives)
...
+                {
+                    return Windows.GetProcAddress(this.handle, symbolName);
+                }
+                else
+                {
+                    // Yes, we could potentially predict the size... but it's a lot simpler to just try
+                    // all the candidates. Most functions have a suffix of @0, @4 or @8 so we won't be trying
+                    // many options - and if it takes a little bit longer to fail if we've really got the wrong
+                    // library, that's not a big problem. This is only called once per function in the native library.
+                    symbolName = "_" + symbolName + "@";
```
Comment (apolcyn) [inline src/csharp/Grpc.Core/Internal/UnmanagedLibrary.cs]: From the stackoverflow post, this works for C __stdcall functions, which are common for 32 bits windows? Not sure if all of the 32bit windows libraries use __stdcall, but would there be C functions that the DLLImport could find that wouldn't match these names?

Comment (jskeet) [inline src/csharp/Grpc.Core/Internal/UnmanagedLibrary.cs]: Put it this way: running `dumpbin /exports grpc_csharp_ext.dll` showed everything being done this way.

Now it could be that if we change how the library is built, that would change what's exported...

(We could try *also* just finding the name without any prefix/suffix, of course.)

Comment (apolcyn) [inline src/csharp/Grpc.Core/Internal/UnmanagedLibrary.cs]: Thanks, saw here https://msdn.microsoft.com/en-us/library/aa984739(v=vs.71).aspx that DLLImport was limited to __std_call anyways, so I think its the same then.

File: src/csharp/Grpc.Core/Grpc.Core.targets [hunk 2]
```diff
@@ -1,29 +1,29 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup>
-    <Content Include="$(MSBuildThisFileDirectory)..\..\build\native\bin\windows_x86\grpc_csharp_ext.dll">
+    <Content Include="$(MSBuildThisFileDirectory)..\..\runtimes\win-x86\native\grpc_csharp_ext.x86.dll">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-      <Link>nativelibs\windows_x86\grpc_csharp_ext.dll</Link>
+      <Link>grpc_csharp_ext.x86.dll</Link>
```
Comment (apolcyn) [inline src/csharp/Grpc.Core/Grpc.Core.targets]: So with this change, all library versions have a unique name, rather than needing to be placed in a unique directory? Is this why this now just refers to the name here?

Comment (jskeet) [inline src/csharp/Grpc.Core/Grpc.Core.targets]: Exactly. This is partly because that's the way the dotnet cli build system seems to do, and partly because it's more flexible in general.

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs [hunk 3]
```diff
@@ -97,21 +94,31 @@ public NativeMethods NativeMethods
...
+            {
+                paths = new[] { explicitPath };
+            }
+            else
+            {
+                var assemblyDirectory = Path.GetDirectoryName(GetAssemblyPath());
+
+                // With old-style VS projects, the native libraries get copied using a .targets rule to the build output folder
+                // alongside the compiled assembly.
+                // With dotnet cli projects, the native libraries (just the required ones) are similarly copied to the built output folder,
```
Comment (apolcyn) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs]: Is this why the the shared library sources are now kept in ".../runtime/<os>-<arch>/..." instead of ".../build/..." directories? Is this following a dotnet convention of where it finds the required library?

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs [hunk 4]
```diff
@@ -97,21 +94,31 @@ public NativeMethods NativeMethods
...
+                // through the magic of Microsoft.NETCore.Platforms.
+                var classicPath = Path.Combine(assemblyDirectory, GetNativeLibraryFilename());
+
+                // DNX-style project.json projects will use Grpc.Core assembly directly in the location where it got restored
+                // by nuget. We locate the native libraries based on known structure of Grpc.Core nuget package.
```
Comment (apolcyn) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs]: So the classic VS and dotnet projects find and copy the library, but these projects using project.json just look them up in the Nuget package and don't copy?

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs (hunk 3 above)
Comment (jskeet) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs]: Yes. I'd love to be able to show you where it's clearly documented, but there's very little in the way of documentation. The closest I've got is https://docs.nuget.org/create/uwp-create

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs (hunk 4 above)
Comment (jskeet) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs]: - Classic VS will end up with *all* the native libraries
- dotnet cli projects will end up with just the right native library alongside all the other assemblies
- dnx projects (which should be considered deprecated now, to be honest - but let's support if we can, for the moment) load the assemblies from the package directory... at least in the cases we've looked at. (Again, it's not terribly clear)

**Comment by apolcyn** [comment]
Not sure why some of the tests look to be unable to load the native library

**Comment by jskeet** [comment]
Will have a look over the weekend, or on Monday. Fingers crossed it's something simple...

**Comment by jskeet** [comment]
I've had a thought - I suspect it's due to where the Grpc.Core csproj files copies the native library - basically the code expects to find it *as it's packaged in nuget*. Fingers crossed that'll be easy to fix...

**Comment by jskeet** [comment]
Okay, that seems to have improved the tests - not sure why the basic and performance tests are failing though. The basic tests doesn't seem to be the C# ones anyway...

**Comment by pgrosu** [comment]
Hi Jon,

To make it easier to read I organized this in two parts, one for each test:

##### The Basic tests

The reason the Basic tests failed is as follows, and mainly because of the `dbg,c,linux` build is failing:

1) I will reference the log which is located here:

https://grpc-testing.appspot.com/job/gRPC_pull_requests/10767/config=dbg,language=c,platform=linux/console

2) The main error happens because the `/home/jenkins/workspace/gRPC_pull_requests/config/dbg/language/c/platform/linux` directory cannot be created created:

```
Building remotely on grpc-jenkins-linux2 (linux) in workspace /home/jenkins/workspace/gRPC_pull_requests/config/dbg/language/c/platform/linux
java.io.IOException: Failed to mkdirs: /home/jenkins/workspace/gRPC_pull_requests/config/dbg/language/c/platform/linux
	at hudson.FilePath.mkdirs(FilePath.java:1191)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1268)
	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:604)
	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:529)
	at hudson.model.Run.execute(Run.java:1720)
	at hudson.matrix.MatrixRun.run(MatrixRun.java:146)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
```

The reason for this is that the directory cannot be created for several reasons:

- The `jenkins` user does not have proper permissions to create a directory somewhere along the path `/home/jenkins/workspace/gRPC_pull_requests/config/dbg/language/c/platform/linux`

- The SSH keys might not be properly set up.

- Or another reason that eventually falls back to one or both of the above two.

I do not have access to the Google Jenkins setup to help troubleshoot, and determine which it might  the precise cause - or if it is something else - but below is a link to a detailed manual for Distributed Builds using Jenkins:

https://wiki.jenkins-ci.org/display/JENKINS/Distributed+builds

##### The Performance smoke test

The reason the Peformance smoke test failed is as follows:

1) I will reference the log which is located here:

https://grpc-testing.appspot.com/job/gRPC_performance_pull_requests/4389/console

2) The performance test is triggered by the following script:

[tools/jenkins/run_performance.sh](https://github.com/grpc/grpc/blob/f484ba9eb64f4d6f8a380da910befd1f6fa86801/tools/jenkins/run_performance.sh#L37)

And the actual line that runs it is this:

```Bash
tools/run_tests/run_performance_tests.py -l c++ node ruby csharp python --netperf --category smoketest
```

If the [run_performance_tests.py](https://github.com/grpc/grpc/blob/f484ba9eb64f4d6f8a380da910befd1f6fa86801/tools/run_tests/run_performance_tests.py) script fails, it will cascade up to Jenkins to return the `Build step 'Execute shell' marked build as failure` status.

3)  The reason why it is failing is because of the following, stemming from the [run_worker_python.sh](https://github.com/grpc/grpc/blob/master/tools/run_tests/performance/run_worker_python.sh) script:

```
++ dirname tools/run_tests/performance/run_worker_python.sh
+ cd tools/run_tests/performance/../../..
+ PYTHONPATH=src/python/grpcio_tests:src/python/gens
+ py27/bin/python src/python/grpcio_tests/tests/qps/qps_worker.py --driver_port=10500
I0709 08:34:59.354176604   21856 ev_epoll_linux.c:84]        epoll engine will be using signal: 36
D0709 08:34:59.354213858   21856 ev_posix.c:106]             Using polling engine: epoll
Traceback (most recent call last):
  File "src/python/grpcio_tests/tests/qps/qps_worker.py", line 61, in <module>
    run_worker_server(args.port)
  File "src/python/grpcio_tests/tests/qps/qps_worker.py", line 43, in run_worker_server
    server = grpc.server((), futures.ThreadPoolExecutor(max_workers=5))
  File "/home/jenkins/workspace/gRPC_performance_pull_requests/py27/local/lib/python2.7/site-packages/grpc/__init__.py", line 1230, in server
    return _server.Server(thread_pool, () if handlers is None else handlers)
  File "/home/jenkins/workspace/gRPC_performance_pull_requests/py27/local/lib/python2.7/site-packages/grpc/_server.py", line 739, in __init__
    completion_queue, server, generic_handlers, thread_pool)
  File "/home/jenkins/workspace/gRPC_performance_pull_requests/py27/local/lib/python2.7/site-packages/grpc/_server.py", line 603, in __init__
    self.generic_handlers = list(generic_handlers)
TypeError: 'ThreadPoolExecutor' object is not iterable
Exception AttributeError: "'Server' object has no attribute '_state'" in <bound method Server.__del__ of <grpc._server.Server object at 0x7f7ee618dcd0>> ignored
```

The reason the above is failing, is because in [line 603 of the src/python/grpcio/grpc/_server.py](https://github.com/grpc/grpc/blob/340d39619df721eee1ebdce9da983a6177207cc5/src/python/grpcio/grpc/_server.py#L603) script, it tries to make `generic_handlers` a list which it cannot, since it is a `ThreadPoolExecutor` object - (the `AttributeError` exception of no attribute '_state' happens in [lines 756-757](https://github.com/grpc/grpc/blob/340d39619df721eee1ebdce9da983a6177207cc5/src/python/grpcio/grpc/_server.py#L756-757), and is inconsequential since the object is not built because of the first error):

```Python
self.generic_handlers = list(generic_handlers)
```

A `ThreadPoolExecutor` is a type of object as noted in the following link, under section 17.4.2:

https://docs.python.org/3/library/concurrent.futures.html

To actually create a list of objects, it might be best to append them as follows:

```Bash
>>> class MyClass:
...     """A simple example class"""
...     i = 12345
...
>>> listOfObjects = []
>>> listOfObjects.append(MyClass)
>>> listOfObjects.append(MyClass)
>>> listOfObjects
[<class __main__.MyClass at 0x7ffff7f0a6b0>, <class __main__.MyClass at 0x7ffff7f0a6b0>]
>>> type(listOfObjects)
<type 'list'>
```

Hope it helps,
Paul

File: src/csharp/Grpc.Core.Tests/project.json [hunk 5]
```diff
@@ -14,10 +14,10 @@
         },
         "copyToOutput": {
           "mappings": {
-            "nativelibs/windows_x64/grpc_csharp_ext.dll": "../../../vsprojects/x64/Debug/grpc_csharp_ext.dll",
-            "nativelibs/windows_x86/grpc_csharp_ext.dll": "../../../vsprojects/Debug/grpc_csharp_ext.dll",
-            "nativelibs/linux_x64/libgrpc_csharp_ext.so": "../../../libs/dbg/libgrpc_csharp_ext.so",
-            "nativelibs/macosx_x64/libgrpc_csharp_ext.dylib": "../../../libs/dbg/libgrpc_csharp_ext.dylib"
+            "grpc_csharp_ext.x64.dll": "../../../vsprojects/x64/Debug/grpc_csharp_ext.dll",
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core.Tests/project.json]: btw, this file is generated using a mako template
(e.g. you need to update the template and run tools/buildgen/generate_project.sh - otherwise the sanity test will fail to indicate that)

Feel free to look at my PR https://github.com/grpc/grpc/pull/7295  (I am updating templates/src/csharp/build_options.include there).

File: src/csharp/Grpc.Core/Grpc.Core.nuspec [hunk 6]
```diff
@@ -25,11 +25,11 @@
...
-    <file src="..\nativelibs\linux_x86\libgrpc_csharp_ext.so" target="/build/native/bin/linux_x86/libgrpc_csharp_ext.so" />
-    <file src="..\nativelibs\linux_x64\libgrpc_csharp_ext.so" target="/build/native/bin/linux_x64/libgrpc_csharp_ext.so" />
-    <file src="..\nativelibs\macosx_x86\libgrpc_csharp_ext.dylib" target="/build/native/bin/macosx_x86/libgrpc_csharp_ext.dylib" />
-    <file src="..\nativelibs\macosx_x64\libgrpc_csharp_ext.dylib" target="/build/native/bin/macosx_x64/libgrpc_csharp_ext.dylib" />
+    <file src="..\nativelibs\windows_x86\grpc_csharp_ext.dll" target="/runtimes/win-x86/native/grpc_csharp_ext.x86.dll" />
+    <file src="..\nativelibs\windows_x64\grpc_csharp_ext.dll" target="/runtimes/win-x64/native/grpc_csharp_ext.x64.dll" />
+    <file src="..\nativelibs\linux_x86\libgrpc_csharp_ext.so" target="/runtimes/linux-x86/native/libgrpc_csharp_ext.x86.so" />
+    <file src="..\nativelibs\linux_x64\libgrpc_csharp_ext.so" target="/runtimes/linux-x64/native/libgrpc_csharp_ext.x64.so" />
+    <file src="..\nativelibs\macosx_x86\libgrpc_csharp_ext.dylib" target="/runtimes/osx-x86/native/libgrpc_csharp_ext.x86.dylib" />
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Grpc.Core.nuspec]: if you put both x86 and x64 libraries under /runtimes/win  /runtimes/linux and /runtimes/osx, it will solve the 32bit vs 64bit runtime issue I've been facing in #7295 (and solved it that way), at the cost of copying both x86 and x64 libraries to the output directory, which I think is a very reasonable tradeoff.

**Comment by jtattermusch** [comment]
Overall, I think the changes in this PR are reasonable. 

C# tests are passing, which is good (but currently they are only running with .NET and mono, not CoreCLR).
it might be worth checking coreCLR tests manually:
`tools/run_tests/run_tests.py -l csharp --compiler coreclr`  (and optionally --use_docker if you don't have dotnet cli installed locally).

We also have tests for distribution packages (the nugets) - but they have to be spawned manually as the build process is a bit more sophisticated:
I've started https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_build_artifacts/1214/ and eventually, nuget packages will be built from the native artifacts built here and they will be tested by the https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_distribtest/ build (I'll post a link there once the previous steps finish).
For dotnet CLI environment, we don't have any distribtests yet (currently only for desktop .NET 45 on windows and mono on Linux and Mac) - so with dotnet CLI the nugets need to be tested manually.

File: src/csharp/Grpc.Core/GrpcEnvironment.cs [hunk 7]
```diff
@@ -180,6 +181,45 @@ public static Task KillServersAsync()
...
+        /// It is highly recommended that the value is an absolute file name rather than a relative one.
+        /// If the file does not exist or is not a valid native library, no further attempt is made to auto-detect
+        /// the correct file; the operation will simply fail.
+        /// </para>
+        /// <para>
+        /// Changes made to this property after the native library is loaded have no effect. This property is
+        /// thread-safe, so may be set from one thread even if the native library ends up being initialized on another.
+        /// </para>
+        /// </remarks>
+        public static string NativeLibrary
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/GrpcEnvironment.cs]: You are not planning to use this property anywhere in this PR, right?
I think it would make sense to add it in a separate PR, as this property is only intended for expert use and for the things we are trying to fix here, no one should be relying on it.

Comment (jtattermusch) [inline src/csharp/Grpc.Core/GrpcEnvironment.cs]: (in case you insist on adding this property here, please mark it as experimental API that can change any time - that's what I've done with other properties I wasn't 100% sure about)

**Comment by jtattermusch** [comment]
Distribtest links:
Nugets for this PR are here:
https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_build_packages/1195/platform=windows/

Experimental nugets with netstandard support are here:
https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_build_packages/1195/platform=linux/artifact/artifacts/

Distrib test results for the "official" nugets are here:
https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_distribtest/1150/  (C# is green, so this PR should be safe).

File: src/csharp/Grpc.Core/GrpcEnvironment.cs (hunk 7 above)
Comment (jskeet) [inline src/csharp/Grpc.Core/GrpcEnvironment.cs]: Will move it to a separate PR, and mark it as experimental as suggested.

File: src/csharp/Grpc.Core.Tests/project.json (hunk 5 above)
Comment (jskeet) [inline src/csharp/Grpc.Core.Tests/project.json]: Thanks - I suspect I haven't got everything I need on my Windows box to run that (as it's failing, in a way which is slightly annoying to debug due bash closing automatically on Windows...) Will look into it.

Comment (pgrosu) [inline src/csharp/Grpc.Core.Tests/project.json]: Hi Jon,

I would just open a `cmd.exe` window, and launch a Bash shell via Cygwin as follows:

```
c:\cygwin\bin\bash.exe --login -i
```

Then proceed to where your Bash script is located, and run it manually as on a normal Linux/Unix box.  The drives on Windows are located under `/cygdrive` like this:

```
$ ls /cygdrive
c
```

This way when running the script, you can see the debug information without having to worry that the window will close.

Hope it helps,
~p

Comment (jskeet) [inline src/csharp/Grpc.Core.Tests/project.json]: Well I don't have cygwin installed and don't wish to go that way - I've made some progress on Linux, although the script is now failing in boringssl. I have other more urgent demands at the moment, but will come back to this when I get a chance.

Comment (pgrosu) [inline src/csharp/Grpc.Core.Tests/project.json]: No problem, we'll be here if you need us.  Not sure if it's the same thing, but I've run into issues with BoringSSL previously, where I had to make a change to the `Makefile` to disable hidden visibility:

```
$ cp Makefile Makefile-orig
$ sed s/-fvisibility=hidden//g Makefile > Makefile-not-hidden
$ cp Makefile-not-hidden Makefile
```

I've posted the whole thing with more details under issue https://github.com/grpc/grpc/issues/7298

~p

Comment (jskeet) [inline src/csharp/Grpc.Core.Tests/project.json]: Thanks for the very quick response - will give that a try while I'm waiting for other build services to probably fail. Today is apparently "swear at infrastructure" day :(

Comment (jskeet) [inline src/csharp/Grpc.Core.Tests/project.json]: Worked it out now - I didn't have `go` installed. Looking better now. I suspect that `CONTRIBUTING.md` could do with being updated for the various dependencies required before running the toolchain. (I didn't have pyyaml or mako installed before today either...)

Comment (pgrosu) [inline src/csharp/Grpc.Core.Tests/project.json]: Great!  Glad to hear it's working out :)

**Comment by apolcyn** [comment]
LGTM

**Comment by jskeet** [comment]
@apolcyn: This isn't the PR that should be merged. See #7401. I'll close this now to avoid confusion :)

**Comment by jskeet** [comment]
(We'll need to merge the changes in #7401 onto master as well, of course.)
